File: base-java/src/main/java/tech/thdev/base/view/BasePresenterFragment.java
package tech.thdev.base.view;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;

import tech.thdev.base.presenter.BasePresenter;
import tech.thdev.base.presenter.BaseView;

/**
 * Created by tae-hwan on 5/6/17.
 */

public abstract class BasePresenterFragment<VIEW extends BaseView, P extends BasePresenter<VIEW>> extends BaseFragment implements BaseView {

    protected P presenter;

    protected abstract P onCreatePresenter();

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        presenter = onCreatePresenter();
        presenter.attachView((VIEW) this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }

    @Override
    public boolean isNotFinish() {
        return !getActivity().isFinishing();
    }
}

File: base-kotlin/src/main/java/tech/thdev/base/view/BasePresenterFragment.kt
package tech.thdev.base.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import tech.thdev.base.presenter.BasePresenter
import tech.thdev.base.presenter.BaseView

/**
 * Created by tae-hwan on 8/28/16.
 */
abstract class BasePresenterFragment<in VIEW : BaseView, PRESENTER : BasePresenter<VIEW>> : BaseFragment(), BaseView {

    protected lateinit var presenter: PRESENTER
        private set

    abstract fun onCreatePresenter(): PRESENTER

    @Suppress("UNCHECKED_CAST")
    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        presenter = onCreatePresenter()
        presenter.attachView(this as VIEW)
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.detachView()
    }
}